### PR TITLE
multi_heap: Introduce shared multi-heap memory pool manager

### PR DESCRIPTION
**Pull request by carlocaione**
The shared multi-heap memory pool manager uses the multi-heap allocator to manage a set of reserved memory regions with different capabilities / attributes (cacheable, non-cacheable, etc...) defined in the DT.

The user can request allocation from the shared pool specifying the capability / attribute of interest for the memory (cacheable / non-cacheable memory, etc...)

For example, in the DT we can have:
```
/ {
        reserved-memory {
                compatible = "reserved-memory";
                #address-cells = <1>;
                #size-cells = <1>;

                res0: reserved@42000000 {
                        compatible = "shared-multi-heap";
                        reg = <0x42000000 0x1000>;
                        capability = "cacheable";
                        label = "res0";
                };

                res1: reserved@43000000 {
                        compatible = "shared-multi-heap";
                        reg = <0x43000000 0x2000>;
                        capability = "non-cacheable";
                        label = "res1";
                };

                res2: reserved2@44000000 {
                        compatible = "shared-multi-heap";
                        reg = <0x44000000 0x3000>;
                        capability = "cacheable";
                        label = "res2";
                };
        };
};
```
And we can request cacheable or non-cacheable heap memory with:
```
shared_multi_heap_alloc(SMH_REG_ATTR_CACHEABLE, 0x1000);
shared_multi_heap_alloc(SMH_REG_ATTR_NON_CACHEABLE, 0x1000);
```

The subsystem is easily extendable with additional attributes.

### CODE REVIEW

**Review by npitre — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]

File: include/memory_pool/shared_multi_heap.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+struct smh_region {
+	enum smh_reg_attr attr;
+	uintptr_t addr;
+	size_t size;
+};
+
+/**
+ * @brief Region init function
+ *
+ * This is a user-provided function whose responsability is to setup or
```
[the remark this answers is not in the document]
Comment (carlocaione) [reply include/memory_pool/shared_multi_heap.h:58]: :( right

**Review by nashif — CHANGES_REQUESTED** [review]
@carlocaione I like this, a few things to consider:
- Is this a subsystem really? IMO this should be side by side with the multiheap, ie. in lib, we might consider moving all of this into one single place for memory management.
- Calling this memory_pool is going to be confusing. We has a memory pool object in the past, we should consider naming this differently so this does not collide with memory pools in previous versions of zephyr.

**Comment by carlocaione** [comment]
> * Is this a subsystem really? IMO this should be side by side with the multiheap, ie. in lib, we might consider moving all of this into one single place for memory management.

Right, the old subsystem vs driver vs lib story :sweat_smile:

I'm not against moving this into `lib` but it would look a bit out of place IMO:
- We are using the `reserved-memory` nodes and in general the whole code is based on information coming from DT. AFAICT this would be the first library in `lib` to use DT helpers. On the contrary a lot of subsystems are using information coming from DT.
- This is a specific implementation of a memory pool manager. In the future someone else could introduce something different using a different compatible with a different backend mechanics and this is exactly what I would expect from a subsystem (less so from a library).

Maybe we should raise the point to the next dev meeting?

> * Calling this memory_pool is going to be confusing. We has a memory pool object in the past, we should consider naming this differently so this does not collide with memory pools in previous versions of zephyr.

I was not aware of this. This work is inspired by the `shared-dma-pool` compatible in Linux, hence the name. Maybe `multi heap pool manager` ?

File: include/memory_pool/shared_multi_heap.h [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * @brief Memory region attributes / capabilities
+ *
+ * !!! This list needs to be kept in sync with shared-multi-heap-pool.yaml !!!
+ */
+enum smh_reg_attr {
+	SMH_REG_ATTR_CACHEABLE,
+	SMH_REG_ATTR_NON_CACHEABLE,
+	SMH_REG_ATTR_NUM,
```
Comment (andyross) [inline include/memory_pool/shared_multi_heap.h:39]: Is there a specific use case for these guys?  Seems like this is going to end up being very platform dependent.  Might want some kind of extension mechanism or something so every platform isn't having to come in and modify the global header.

Again, the use case I'm looking at are DSPs with three different memory regions of widely varying latencies (~12, 30 and 100+ cycles!), different power properties (one is low power always, but small, one has 64k banks that can be shut off programmatically), translation capability (one has a primitive global MMU, the others are direct mapped).  And each of those is actually mapped twice, once behind a per-CPU L1 cache and once with synchronized uncached access.  I despair of a general solution ever working for that, and even if it did no one would want to use that monster for simpler platforms.

tl;dr: This probably won't work for the audio DSP stuff, but that's a weird edge case.  Are we sure this is generic enough for "normal" apps?

File: include/memory_pool/shared_multi_heap.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+/**
+ * @brief Init the pool
+ *
+ * Initialize the shared multi-heap memory pool and hook-up the region init
+ * function.
+ *
+ * @param smh_init_reg_fn The function pointer to the region init function. Can
+ *                        be NULL for non-MPU / non-MMU architectures.
+ */
+int smh_pool_init(smh_init_reg_fn_t smh_init_reg_fn);
```
Comment (andyross) [inline include/memory_pool/shared_multi_heap.h:85]: If it's a single global device instantiated out of devicetree, why bother exposing the initialization hook in the public API?  Seems like it could just live as a SYS_INIT() callback internally.

**Review by andyross — APPROVED** [review]
Looks great on the whole, though I have one worry about how applicable this will be to generic problems.

I take no position on the "library vs. subsystem debate".  And it's true, our memory manager code used to be called "mem_pool" (there were two attempts with that name) and it had a bunch of issues, including some API weirdness that made it hard to support a generic malloc/free interface at a low level without waste.  The new code was renamed "heap" to distinguish it and allow it to coexist alongside the slightly-incompatible older version for a release.

As to whether we should re-use the name or not to conform with other practice in the industry, again I won't engage. :)

**Comment by carlescufi** [comment]
> @carlocaione I like this, a few things to consider:
> 
> * Is this a subsystem really? IMO this should be side by side with the multiheap, ie. in lib, we might consider moving all of this into one single place for memory management.

I agree with @nashif  here. I suggest we place this in `lib/os` for the time being, until it becomes necessary to create its own folder.

> * Calling this memory_pool is going to be confusing. We has a memory pool object in the past, we should consider naming this differently so this does not collide with memory pools in previous versions of zephyr.

I also agree with @nashif here. I don't have a good alternative but `memory pool` is a loaded term in Zephyr at this point.

File: include/memory_pool/shared_multi_heap.h (hunk 3 above)
Comment (carlocaione) [reply include/memory_pool/shared_multi_heap.h:85]: But how do you pass the `smh_init_reg_fn` function to the framework in that case? `smh_init_reg_fn` is a user-provided function that is used as callback for the memory regions initialization.

**Comment by carlocaione** [comment]
> I agree with @nashif here.

Fair enough, I'll move this into `lib` and I'll change the name.

File: include/memory_pool/shared_multi_heap.h (hunk 2 above)
Comment (carlocaione) [reply include/memory_pool/shared_multi_heap.h:39]: > Is there a specific use case for these guys?

Yup, there is. Having the having different memory regions with different cache-ability attributes is not uncommon, especially when you have multiple cores, each one behind a different cache for example.

But without going into more complex scenarios with multiple regions, I think this code is also handy also when for example you need to have one "no-cache" area for DMA where to take memory from. This is currently done in some drivers by leveraging the `CONFIG_NOCACHE_MEMORY` and some well-placed linker script symbols (see for example https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/dma/loop_transfer/src/test_dma_loop.c#L36 or https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/device/usb_dc_mcux.c#L69). So in general is not uncommon to need heap allocation based on cacheability.

With this PR all the information is in the DT with a nice API on top to retrieve the needed memory heap.

> Might want some kind of extension mechanism or something so every platform isn't having to come in and modify the global header.

Yeah, I thought about that, but that would shift even more complexity and code from the pool manager to the user code and at least for now I'd like to keep this as simple as possible.

> Are we sure this is generic enough for "normal" apps?

I mean, it's probably not something that the wider audience will ever find useful but for sure this serves well for some more complex cases.

**Comment by nashif** [comment]
@carlocaione thank you for making the changes. Can you please add some documentation to that as well?

**Review by nika-nordic — APPROVED** [review]

**Comment by carlocaione** [comment]
@nashif documentation added

**Review by nashif — APPROVED** [review]

**Review by mbolivar-nordic — COMMENTED** [review]
Late to the party, but the DT binding looks fine.
